man/man3/drawfcall.3
.TH DRAWFCALL 3
.SH NAME
convM2W, convW2M, drawfcallfmt, readwsysmsg, sizeW2M \- window system protocol
.SH SYNOPSIS
.nf
.PP
.ft L
#include <u.h>
#include <libc.h>
#include <draw.h>
#include <mouse.h>
#include <cursor.h>
#include <drawfcall.h>
.ft P
.ta \w'\fLuint 'u
.PP
.B
uint	convM2W(uchar *ap, uint nap, Wsysmsg *w)
.PP
.B
uint	convW2M(Wsysmsg *w, uchar *ap, uint nap)
.PP
.B
int	drawfcallfmt(Fmt*)
.PP
.B
int	readwsysmsg(int fd, uchar *buf, uint nbuf)
.PP
.B
uint	sizeW2M(Wsysmsg *w)
.SH DESCRIPTION
These routines are analogues of the routines described in 
.MR fcall 3 .
They manipulate graphics device protocol messages
rather than 9P protocol messages.
The graphics device protocol is used for internal 
communication between the
.MR devdraw 1
graphics server
and the
.MR draw 3
library.
A
.B Wsysmsg
is the C structure corresponding to a protocol message.
.PP
The protocol is intentionally undocumented and may change.
.SH SOURCE
.B \*9/src/libdraw/drawfcall.c
.SH SEE ALSO
.MR devdraw 1 ,
.MR draw 3 ,
.MR graphics 3


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/libdraw/drawfcall.c
#include <u.h>
#include <libc.h>
#include <draw.h>
#include <mouse.h>
#include <cursor.h>
#include <drawfcall.h>

static int
_stringsize(char *s)
{
	if(s == nil)
		s = "";
	return 4+strlen(s);
}

static int
PUTSTRING(uchar *p, char *s)
{
	int n;

	if(s == nil)
		s = "";
	n = strlen(s);
	PUT(p, n);
	memmove(p+4, s, n);
	return n+4;
}

static int
GETSTRING(uchar *p, char **s)
{
	int n;

	GET(p, n);
	memmove(p, p+4, n);
	*s = (char*)p;
	p[n] = 0;
	return n+4;
}

uint
sizeW2M(Wsysmsg *m)
{
	switch(m->type){
	default:
		return 0;
	case Trdmouse:
	case Rbouncemouse:
	case Rmoveto:
	case Rcursor:
	case Rcursor2:
	case Trdkbd:
	case Trdkbd4:
	case Rlabel:
	case Rmode:
	case Rctxt:
	case Rinit:
	case Trdsnarf:
	case Rwrsnarf:
	case Ttop:
	case Rtop:
	case Rresize:
		return 4+1+1;
	case Rrdmouse:
		return 4+1+1+4+4+4+4+1;
	case Tbouncemouse:
		return 4+1+1+4+4+4;
	case Tmoveto:
		return 4+1+1+4+4;
	case Tcursor:
		return 4+1+1+4+4+2*16+2*16+1;
	case Tcursor2:
		return 4+1+1+4+4+2*16+2*16+4+4+4*32+4*32+1;
	case Rerror:
		return 4+1+1+_stringsize(m->error);
	case Rrdkbd:
		return 4+1+1+2;
	case Rrdkbd4:
		return 4+1+1+4;
	case Tlabel:
		return 4+1+1+_stringsize(m->label);
	case Tmode:
		return 4+1+1+4;
	case Tctxt:
		return 4+1+1
			+_stringsize(m->id);
	case Tinit:
		return 4+1+1
			+_stringsize(m->winsize)
			+_stringsize(m->label);
	case Rrdsnarf:
	case Twrsnarf:
		return 4+1+1+_stringsize(m->snarf);
	case Rrddraw:
	case Twrdraw:
		return 4+1+1+4+m->count;
	case Trddraw:
	case Rwrdraw:
		return 4+1+1+4;
	case Tresize:
		return 4+1+1+4*4;
	}
}

uint
convW2M(Wsysmsg *m, uchar *p, uint n)
{
	int nn;

	nn = sizeW2M(m);
	if(n < nn || nn == 0 || n < 6)
		return 0;
	PUT(p, nn);
	p[4] = m->tag;
	p[5] = m->type;

	switch(m->type){
	default:
		return 0;
	case Trdmouse:
	case Rbouncemouse:
	case Rmoveto:
	case Rcursor:
	case Rcursor2:
	case Trdkbd:
	case Trdkbd4:
	case Rlabel:
	case Rmode:
	case Rctxt:
	case Rinit:
	case Trdsnarf:
	case Rwrsnarf:
	case Ttop:
	case Rtop:
	case Rresize:
		break;
	case Rerror:
		PUTSTRING(p+6, m->error);
		break;
	case Rrdmouse:
		PUT(p+6, m->mouse.xy.x);
		PUT(p+10, m->mouse.xy.y);
		PUT(p+14, m->mouse.buttons);
		PUT(p+18, m->mouse.msec);
		p[19] = m->resized;
		break;
	case Tbouncemouse:
		PUT(p+6, m->mouse.xy.x);
		PUT(p+10, m->mouse.xy.y);
		PUT(p+14, m->mouse.buttons);
		break;
	case Tmoveto:
		PUT(p+6, m->mouse.xy.x);
		PUT(p+10, m->mouse.xy.y);
		break;
	case Tcursor:
		PUT(p+6, m->cursor.offset.x);
		PUT(p+10, m->cursor.offset.y);
		memmove(p+14, m->cursor.clr, sizeof m->cursor.clr);
		memmove(p+46, m->cursor.set, sizeof m->cursor.set);
		p[78] = m->arrowcursor;
		break;
	case Tcursor2:
		PUT(p+6, m->cursor.offset.x);
		PUT(p+10, m->cursor.offset.y);
		memmove(p+14, m->cursor.clr, sizeof m->cursor.clr);
		memmove(p+46, m->cursor.set, sizeof m->cursor.set);
		PUT(p+78, m->cursor2.offset.x);
		PUT(p+82, m->cursor2.offset.y);
		memmove(p+86, m->cursor2.clr, sizeof m->cursor2.clr);
		memmove(p+214, m->cursor2.set, sizeof m->cursor2.set);
		p[342] = m->arrowcursor;
		break;
	case Rrdkbd:
		PUT2(p+6, m->rune);
		break;
	case Rrdkbd4:
		PUT(p+6, m->rune);
		break;
	case Tlabel:
		PUTSTRING(p+6, m->label);
		break;
	case Tmode:
		PUT(p+6, m->mode);
		break;
	case Tctxt:
		PUTSTRING(p+6, m->id);
		break;
	case Tinit:
		p += 6;
		p += PUTSTRING(p, m->winsize);
		p += PUTSTRING(p, m->label);
		break;
	case Rrdsnarf:
	case Twrsnarf:
		PUTSTRING(p+6, m->snarf);
		break;
	case Rrddraw:
	case Twrdraw:
		PUT(p+6, m->count);
		memmove(p+10, m->data, m->count);
		break;
	case Trddraw:
	case Rwrdraw:
		PUT(p+6, m->count);
		break;
	case Tresize:
		PUT(p+6, m->rect.min.x);
		PUT(p+10, m->rect.min.y);
		PUT(p+14, m->rect.max.x);
		PUT(p+18, m->rect.max.y);
		break;
	}
	return nn;
}

uint
convM2W(uchar *p, uint n, Wsysmsg *m)
{
	int nn;

	if(n < 6)
		return 0;
	GET(p, nn);
	if(nn > n)
		return 0;
	m->tag = p[4];
	m->type = p[5];
	switch(m->type){
	default:
		return 0;
	case Trdmouse:
	case Rbouncemouse:
	case Rmoveto:
	case Rcursor:
	case Rcursor2:
	case Trdkbd:
	case Trdkbd4:
	case Rlabel:
	case Rmode:
	case Rctxt:
	case Rinit:
	case Trdsnarf:
	case Rwrsnarf:
	case Ttop:
	case Rtop:
	case Rresize:
		break;
	case Rerror:
		GETSTRING(p+6, &m->error);
		break;
	case Rrdmouse:
		GET(p+6, m->mouse.xy.x);
		GET(p+10, m->mouse.xy.y);
		GET(p+14, m->mouse.buttons);
		GET(p+18, m->mouse.msec);
		m->resized = p[19];
		break;
	case Tbouncemouse:
		GET(p+6, m->mouse.xy.x);
		GET(p+10, m->mouse.xy.y);
		GET(p+14, m->mouse.buttons);
		break;
	case Tmoveto:
		GET(p+6, m->mouse.xy.x);
		GET(p+10, m->mouse.xy.y);
		break;
	case Tcursor:
		GET(p+6, m->cursor.offset.x);
		GET(p+10, m->cursor.offset.y);
		memmove(m->cursor.clr, p+14, sizeof m->cursor.clr);
		memmove(m->cursor.set, p+46, sizeof m->cursor.set);
		m->arrowcursor = p[78];
		break;
	case Tcursor2:
		GET(p+6, m->cursor.offset.x);
		GET(p+10, m->cursor.offset.y);
		memmove(m->cursor.clr, p+14, sizeof m->cursor.clr);
		memmove(m->cursor.set, p+46, sizeof m->cursor.set);
		GET(p+78, m->cursor2.offset.x);
		GET(p+82, m->cursor2.offset.y);
		memmove(m->cursor2.clr, p+86, sizeof m->cursor2.clr);
		memmove(m->cursor2.set, p+214, sizeof m->cursor2.set);
		m->arrowcursor = p[342];
		break;
	case Rrdkbd:
		GET2(p+6, m->rune);
		break;
	case Rrdkbd4:
		GET(p+6, m->rune);
		break;
	case Tlabel:
		GETSTRING(p+6, &m->label);
		break;
	case Tmode:
		GET(p+6, m->mode);
		break;
	case Tctxt:
		GETSTRING(p+6, &m->id);
		break;
	case Tinit:
		p += 6;
		p += GETSTRING(p, &m->winsize);
		p += GETSTRING(p, &m->label);
		break;
	case Rrdsnarf:
	case Twrsnarf:
		GETSTRING(p+6, &m->snarf);
		break;
	case Rrddraw:
	case Twrdraw:
		GET(p+6, m->count);
		m->data = p+10;
		break;
	case Trddraw:
	case Rwrdraw:
		GET(p+6, m->count);
		break;
	case Tresize:
		GET(p+6, m->rect.min.x);
		GET(p+10, m->rect.min.y);
		GET(p+14, m->rect.max.x);
		GET(p+18, m->rect.max.y);
		break;
	}
	return nn;
}

int
readwsysmsg(int fd, uchar *buf, uint nbuf)
{
	int n;

	if(nbuf < 6)
		return -1;
	if(readn(fd, buf, 4) != 4)
		return -1;
	GET(buf, n);
	if(n > nbuf)
		return -1;
	if(readn(fd, buf+4, n-4) != n-4)
		return -1;
	return n;
}

int
drawfcallfmt(Fmt *fmt)
{
	Wsysmsg *m;

	m = va_arg(fmt->args, Wsysmsg*);
	fmtprint(fmt, "tag=%d ", m->tag);
	switch(m->type){
	default:
		return fmtprint(fmt, "unknown msg %d", m->type);
	case Rerror:
		return fmtprint(fmt, "Rerror error='%s'", m->error);
	case Trdmouse:
		return fmtprint(fmt, "Trdmouse");
	case Rrdmouse:
		return fmtprint(fmt, "Rrdmouse x=%d y=%d buttons=%d msec=%d resized=%d",
			m->mouse.xy.x, m->mouse.xy.y,
			m->mouse.buttons, m->mouse.msec, m->resized);
	case Tbouncemouse:
		return fmtprint(fmt, "Tbouncemouse x=%d y=%d buttons=%d",
			m->mouse.xy.x, m->mouse.xy.y, m->mouse.buttons);
	case Rbouncemouse:
		return fmtprint(fmt, "Rbouncemouse");
	case Tmoveto:
		return fmtprint(fmt, "Tmoveto x=%d y=%d", m->mouse.xy.x, m->mouse.xy.y);
	case Rmoveto:
		return fmtprint(fmt, "Rmoveto");
	case Tcursor:
		return fmtprint(fmt, "Tcursor arrow=%d", m->arrowcursor);
	case Tcursor2:
		return fmtprint(fmt, "Tcursor2 arrow=%d", m->arrowcursor);
	case Rcursor:
		return fmtprint(fmt, "Rcursor");
	case Rcursor2:
		return fmtprint(fmt, "Rcursor2");
	case Trdkbd:
		return fmtprint(fmt, "Trdkbd");
	case Rrdkbd:
		return fmtprint(fmt, "Rrdkbd rune=%C", m->rune);
	case Trdkbd4:
		return fmtprint(fmt, "Trdkbd4");
	case Rrdkbd4:
		return fmtprint(fmt, "Rrdkbd4 rune=%C", m->rune);
	case Tlabel:
		return fmtprint(fmt, "Tlabel label='%s'", m->label);
	case Rlabel:
		return fmtprint(fmt, "Rlabel");
	case Tmode:
		return fmtprint(fmt, "Tmode mode=%d", m->mode);
	case Rmode:
		return fmtprint(fmt, "Rmode");
	case Tctxt:
		return fmtprint(fmt, "Tctxt id='%s'", m->id);
	case Rctxt:
		return fmtprint(fmt, "Rctxt");
	case Tinit:
		return fmtprint(fmt, "Tinit label='%s' winsize='%s'", m->label, m->winsize);
	case Rinit:
		return fmtprint(fmt, "Rinit");
	case Trdsnarf:
		return fmtprint(fmt, "Trdsnarf");
	case Rrdsnarf:
		return fmtprint(fmt, "Rrdsnarf snarf='%s'", m->snarf);
	case Twrsnarf:
		return fmtprint(fmt, "Twrsnarf snarf='%s'", m->snarf);
	case Rwrsnarf:
		return fmtprint(fmt, "Rwrsnarf");
	case Trddraw:
		return fmtprint(fmt, "Trddraw %d", m->count);
	case Rrddraw:
		return fmtprint(fmt, "Rrddraw %d %.*H", m->count, m->count, m->data);
	case Twrdraw:
		return fmtprint(fmt, "Twrdraw %d %.*H", m->count, m->count, m->data);
	case Rwrdraw:
		return fmtprint(fmt, "Rwrdraw %d", m->count);
	case Ttop:
		return fmtprint(fmt, "Ttop");
	case Rtop:
		return fmtprint(fmt, "Rtop");
	case Tresize:
		return fmtprint(fmt, "Tresize %R", m->rect);
	case Rresize:
		return fmtprint(fmt, "Rresize");
	}
}
